### PR TITLE
Enable xlint for build

### LINKED ISSUE #11858: Lint please, hold the deprecation
```
➜  scala git:(topic/linting) scala -deprecation
Welcome to Scala 2.13.1 (OpenJDK 64-Bit Server VM, Java 11.0.3).
Type in expressions for evaluation. Or try :help.

scala> def f: reflect.ClassManifest[_] = ???
                      ^
       warning: type ClassManifest in package reflect is deprecated (since 2.10.0): use scala.reflect.ClassTag instead
f: reflect.ClassManifest[_]

scala> :quit
➜  scala git:(topic/linting) scala -Xlint:-deprecation
Welcome to Scala 2.13.1 (OpenJDK 64-Bit Server VM, Java 11.0.3).
Type in expressions for evaluation. Or try :help.

scala> def f: reflect.ClassManifest[_] = ???
                      ^
       warning: type ClassManifest in package reflect is deprecated (since 2.10.0): use scala.reflect.ClassTag instead
f: reflect.ClassManifest[_]

scala> :quit
➜  scala git:(topic/linting) scala -deprecation:false -Xlint             
Welcome to Scala 2.13.1 (OpenJDK 64-Bit Server VM, Java 11.0.3).
Type in expressions for evaluation. Or try :help.

scala> def f: reflect.ClassManifest[_] = ???
                      ^
       warning: type ClassManifest in package reflect is deprecated (since 2.10.0): use scala.reflect.ClassTag instead
f: reflect.ClassManifest[_]

scala> :quit
➜  scala git:(topic/linting) scala -deprecation:false       
Welcome to Scala 2.13.1 (OpenJDK 64-Bit Server VM, Java 11.0.3).
Type in expressions for evaluation. Or try :help.

scala> def f: reflect.ClassManifest[_] = ???
f: reflect.ClassManifest[_]
```

### PR DESCRIPTION
**Pull request by som-snytt**
Fixes scala/bug#11858

### CODE REVIEW

**Comment by som-snytt** [comment]
Need to exclude flags for partest. Probably if flags are supplied by the test, then don't add any; or maybe just never add behavior flags.

**Comment by dwijnand** [comment]
Probably good to capture in the tests how the various option combinations mentioned in scala/bug#11858 behave.

**Comment by som-snytt** [comment]
Local `testAll` just failed because `test/it` wants to pass `-Ymacro-annotations`, also `compiler/compile:doc` is unhappy. There might be a way to install a global flags file? or was that only for filters?

File: src/partest/scala/tools/partest/nest/DirectCompiler.scala [hunk 1]
```diff
@@ -75,7 +75,7 @@ class DirectCompiler(val runner: Runner) {
     val Xplugin = if (xplugs.isEmpty) Nil else List(xprefix +
       (xplugs map (_ stripPrefix xprefix) flatMap (_ split pathSeparator) map absolutize mkString pathSeparator)
     )
-    runner.suiteRunner.scalacExtraArgs ++ runner.suiteRunner.scalacOpts.split(' ') ++ others ++ Xplugin
+    runner.suiteRunner.scalacExtraArgs ++ /*runner.suiteRunner.scalacOpts.split(' ') ++*/ others ++ Xplugin
```
Comment (SethTisue) [inline src/partest/scala/tools/partest/nest/DirectCompiler.scala]: this should be cleaned up before merge

**Review by SethTisue — CHANGES_REQUESTED** [review]
in build.sbt, the flags are a little different on the different projects. is there an opportunity here to put a few TODO comments for future warnings-slayers? like if you wanted to enable a flag but just didn't get to it in this round?

otherwise LGTM (but, one cleanup request)

File: src/partest/scala/tools/partest/nest/DirectCompiler.scala (hunk 1 above)
Comment (som-snytt) [reply src/partest/scala/tools/partest/nest/DirectCompiler.scala]: It was documentary; but I'm not sure of the right thing to do. 

For example, when building with optimizer turned up, should tests be compiled that way? I know that some of the check files (such as output of -Xshow-phases) differs for that reason. My suspicion is that all test flags should be explicit. But I didn't track down what ExtraArgs is for.

**Comment by som-snytt** [comment]
There isn't a great way to selectively disable lint options in subprojects; or did I miss it?

The unit test seems to say that turning on an option always wins, except maybe for expanding options.

**Comment by SethTisue** [comment]
you know better than me — if you don't see an opportunity here, you can ignore my remark

**Comment by som-snytt** [comment]
I needed this enabled today.

File: src/partest/scala/tools/partest/nest/DirectCompiler.scala (hunk 1 above)
Comment (lrytz) [reply src/partest/scala/tools/partest/nest/DirectCompiler.scala]: Seems this is actually in use? https://github.com/scala/scala/blob/67580388f35f200a1eb0d8ce424556bf5758b596/build.sbt#L1008

Disable this instead? https://github.com/scala/scala/blob/67580388f35f200a1eb0d8ce424556bf5758b596/build.sbt#L864

I agree it's weird to change the test flags depending on the compilation flags. Given that the optimizer is on by default on CI and for releases, enabling it for partest (unconditionally) might be the right thing to do?

Comment (som-snytt) [reply src/partest/scala/tools/partest/nest/DirectCompiler.scala]: The macro-annotations case is handled by the flags file. 

You're suggesting always optimise [sic] for tests? I hadn't given any thought yet as to why that might be useful. Are we talking inline from ** kind of optimizer?

Obviously, I've been working on the -opt flags.

Comment (lrytz) [reply src/partest/scala/tools/partest/nest/DirectCompiler.scala]: > why that might be useful

As an integration test for the optimizer, that's what I can think of.

> inline from ** kind of optimizer

Yes (that's how partest currently runs in CI, I assume?)

Comment (som-snytt) [reply src/partest/scala/tools/partest/nest/DirectCompiler.scala]: OK, so I'll remove the now-extraneous macro flag in the build, and filter the flags in partest to accept opt opts.

**Comment by som-snytt** [comment]
Jenkins is like, Did you even sign the CLA? and I'm Jenkins old chap, you know me! and then Jenkins just messin witcha.

**Comment by som-snytt** [comment]
I'll attempt a test as @dwijnand suggested. FTR, as I've been reminded recently, enabling an option trumps disabling it, except for "expanding" options, in which case explicitly disabling trumps the expanding option. So you enable expanding option `-opt:l:method` but disable one of those flags individually, I'd have to look up what they're called, maybe only Lukas knows.

**Review by lrytz — APPROVED** [review]

**Pull request merged**